### PR TITLE
[JENKINS-58501] add exceptions for metaprogramming to CPS mismatches

### PR DESCRIPTION
**Pull request by steven-terrana**
Hey, thought i'd take a stab at implementing the fix for the pull requests against https://github.com/jenkinsci/workflow-cps-plugin/pull/305 and https://github.com/jenkinsci/workflow-cps-plugin/pull/306.

Just trying to lend a helping hand!  The unit tests added in the workflow-cps-plugin that are failing due to the false positives are passing with this change.

### CODE REVIEW

**Comment by steven-terrana** [comment]
@jglick

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsCallableInvocation.java [hunk 1]
```diff
@@ -57,10 +58,21 @@ public Next invoke(Env caller, SourceLocation loc, Continuation k) {
      * @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-31314">JENKINS-31314</a>
      */
     void checkMismatch(Object expectedReceiver, List<String> expectedMethodNames) {
+        String expectedMethodName = expectedMethodNames.get(0);
+        // metaclass invokeMethod
+        if(MetaClass.class.isInstance(expectedReceiver) && expectedMethodName.equals("invokeMethod")){
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsCallableInvocation.java]: KISS :-)

```suggestion
        if (expectedReceiver instanceof MetaClass && expectedMethodName.equals("invokeMethod")) {
```

**Comment by steven-terrana** [comment]
change committed.  appreciate the fast responses on this!  let me know if there's anything else i can do.

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsCallableInvocation.java [hunk 2]
```diff
@@ -57,10 +58,21 @@ public Next invoke(Env caller, SourceLocation loc, Continuation k) {
      * @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-31314">JENKINS-31314</a>
      */
     void checkMismatch(Object expectedReceiver, List<String> expectedMethodNames) {
+        String expectedMethodName = expectedMethodNames.get(0);
+        // metaclass invokeMethod
+        if (expectedReceiver instanceof MetaClass && expectedMethodName.equals("invokeMethod")) {
```
Comment (dwnusbaum) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsCallableInvocation.java]: Maybe possible to make this work by tweaking `ContinuationGroup.methodCall` and `ContinuationGroup.findEffectiveReceiver` to handle this case, but probably not worth it.

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsCallableInvocation.java [hunk 3]
```diff
@@ -57,10 +58,21 @@ public Next invoke(Env caller, SourceLocation loc, Continuation k) {
      * @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-31314">JENKINS-31314</a>
      */
     void checkMismatch(Object expectedReceiver, List<String> expectedMethodNames) {
+        String expectedMethodName = expectedMethodNames.get(0);
+        // metaclass invokeMethod
```
Comment (dwnusbaum) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsCallableInvocation.java]: nit: This comment doesn't really add any more info or context than the code itself, so I'd delete it.

**Comment by steven-terrana** [comment]
@dwnusbaum resolved the unnecessary comments :). 

Please let me know if there's anything i can do.

**Comment by steven-terrana** [comment]
any update on this?

**Pull request merged**